hack/cloudfront/readme.md

```diff
@@ -0,0 +1,278 @@
+
+If the S3 bucket in this architecture is running in us-east-1 and the client is running in us-east-1 in a standard 
+OpenShift install, this transfer of data is completely free. The problem with this solution is that app.ci
+and all build farms reside in us-east-1, but not all clients are in us-east-1. For example, we stand up
+ephemeral clusters for testing in us-west-2 or GCP. When the client is not in the same region as the S3 bucket, transfer
+costs mount quickly. Transferring content to us-west-2 from us-east-1 costs 0.02 USD. Transferring data to GCP
+(which is just considered "the Internet" to AWS) costs between 0.05-0.09 USD. When you are transferring TB of data,
+this accumulates to a large monthly expense.
+
+A partial solution to this is using CloudFront. AWS CloudFront can be enabled trivially in the OpenShift registry operator's
+CRD. CloudFront can act as a CDN in front of the normal registry S3 bucket to more scalably transfer data from the bucket. 
+The standard pricing for CloudFront is just as expensive (or more so) than just transferring from S3. However, we have a discount
+which reduces the standard cost. 
+
+So if we turn on CloudFront for our registries, transfers to GCP, for example, drop in cost. However, CloudFront will
+also happily serve clients in us-east-1. This means our previously free transfers now incur charges.
+
+# 1000 Foot View
+CloudFront allows serverless functions to be [executed before processing a request itself](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-cloudfront-trigger-events.html). The serverless function can
+modify the request or even respond to the client directly. This means we can write a request handler capable of detecting
+the AWS region from which the client is making a request and redirecting it directly to S3 if we determine, for example,
+that they are an EC2 instance in us-east-1.
+
+So a client from us-east-1 requesting a blob would be processed as follows:
+1. Make a GET request to registry.ci.openshift.org (`oc whoami -t` as the Authorization token).
+2. The registry pod would verify the token and create a presigned URL for the configured CloudFront distribution and send a redirect for the client to the CloudFront URL.
+3. The client receives the redirect and sends a GET to CloudFront. CloudFront verifies the presigned URL.
+4. Our serverless function is invoked by CloudFront and detects that the client IP is from us-east-1.
+5. Our serverless function creates a presigned URL for the registry's S3 storage bucket. It returns the presigned URL to the client as an HTTP redirect.
+6. The client receives the redirect and sends a GET to the S3 HTTP service.
+7. S3 validates the presigned URL and streams the content back to the client.
+Cost: Free except for the low cost of the CloudFront GET request handling and serverless function execution time.
+
+A client from GCP requesting a blob would be processed as follows:
```

Comment:
Is a client from a region other than `us-east-1` a similar case to this one?

Reply:
No. The `100 Foot View` digs into why. Just building on basics in this section.